typescript/sdk/src/ccip/HyperlaneCCIPDeployer.ts

```diff
@@ -0,0 +1,246 @@
+import { CCIPHook__factory, CCIPIsm__factory } from '@hyperlane-xyz/core';
+import {
+  Address,
+  ZERO_ADDRESS_HEX_32,
+  addressToBytes32,
+  assert,
+  rootLogger,
+} from '@hyperlane-xyz/utils';
+
+import {
+  HyperlaneAddressesMap,
+  HyperlaneContracts,
+} from '../contracts/types.js';
+import { CoreAddresses } from '../core/contracts.js';
+import { HyperlaneDeployer } from '../deploy/HyperlaneDeployer.js';
+import { ContractVerifier } from '../deploy/verify/ContractVerifier.js';
+import { MultiProvider } from '../providers/MultiProvider.js';
+import { ChainMap, ChainName } from '../types.js';
+
+import {
+  CCIPContractCache,
+  getCCIPChainSelector,
+  getCCIPRouterAddress,
+} from './utils.js';
+
+export class HyperlaneCCIPDeployer extends HyperlaneDeployer<
+  Set<ChainName>,
+  {}
+> {
+  private ccipContractCache: CCIPContractCache = new CCIPContractCache();
+
+  constructor(
+    multiProvider: MultiProvider,
+    readonly core: ChainMap<Partial<CoreAddresses>>,
+    contractVerifier?: ContractVerifier,
+  ) {
+    super(
+      multiProvider,
+      {},
+      {
+        logger: rootLogger.child({ module: 'HyperlaneCCIPDeployer' }),
+        contractVerifier,
+      },
+    );
+  }
+
+  cacheAddressesMap(addressesMap: HyperlaneAddressesMap<any>): void {
+    super.cacheAddressesMap(addressesMap);
+    this.ccipContractCache.cacheAddressesMap(addressesMap);
+  }
+
+  async deployContracts(
+    origin: ChainName,
+    config: Set<ChainName>,
+  ): Promise<HyperlaneContracts<{}>> {
+    // Deploy ISMs from chain to eachdestination chain concurrently
+    await Promise.all(
+      Array.from(config).map(async (destination) => {
+        // Deploy CCIP ISM for this origin->destination pair
+        await this.deployCCIPIsm(origin, destination);
+      }),
+    );
+
+    //On the origin chain, deploy hooks for each destination chain in series
+    for (const destination of config) {
+      // Grab the ISM from the cache
+      const ccipIsmAddress = this.ccipContractCache.getIsm(origin, destination);
+      assert(
+        ccipIsmAddress,
+        `CCIP ISM not found for ${origin} -> ${destination}`,
+      );
+
+      await this.deployCCIPHook(origin, destination, ccipIsmAddress);
```

Comment:
Why not make this part of the promise above?, i.e. why split this up over multiple loops of destinations?